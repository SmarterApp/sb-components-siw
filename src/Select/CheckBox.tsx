import * as React from "react";
import "../Assets/Styles/multi-select.less";
import { MultiSelectValue } from "@src/index";

export interface MultiSelectProps {
  option: MultiSelectValue;
  onValueChange: (option: MultiSelectValue) => void;
  uniqueId: number;
}

export interface MultiSelectState {
  isChecked: boolean;
}

export class CheckBox extends React.Component<
  MultiSelectProps,
  MultiSelectState
> {
  constructor(props: MultiSelectProps) {
    super(props);
    this.state = {
      isChecked: this.props.option.selected
    };
  }

  handleUpdateOption = () => {
    console.log("Key down code");
    console.log("Before updating state ", this.props.option.selected);
    this.props.option.selected = this.props.option.selected ? false : true;

    this.setState({ isChecked: this.props.option.selected ? false : true });
    console.log("After updating state ", this.props.option.selected);
    this.props.onValueChange(this.props.option);
  };

  handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    console.log("Key down");
    if (e.keyCode === 13 || e.keyCode === 32) {
      e.stopPropagation();
      e.preventDefault();
      if (
        this.props.option.disabled !== undefined &&
        this.props.option.disabled === true
      )
        return;
      this.handleUpdateOption();
    }
  };

  handleCheckBoxChange = (
    e: React.SyntheticEvent,
    option: MultiSelectValue
  ) => {
    e.stopPropagation();
    e.preventDefault();
    if (
      this.props.option.disabled !== undefined &&
      this.props.option.disabled === true
    )
      return;
    this.handleUpdateOption();
  };

  ariaLabelText = () => {
    let ariaLabel: string;
    if (
      this.props.option.disabled !== undefined &&
      this.props.option.disabled === true
    )
      ariaLabel = this.props.option.label + " disabled";
    else ariaLabel = this.props.option.label;
    return ariaLabel;
  };

  shouldDisabled = () => {
    if (
      this.props.option.disabled !== undefined &&
      this.props.option.disabled === true
    )
      return true;
    else return false;
  };

  render() {
    return (
      <label
        style={{ textAlign: "left" }}
        tabIndex={-1}
        aria-disabled={this.shouldDisabled() ? true : false}
        role="checkbox"
        aria-checked={this.props.option.selected}
        aria-label={this.ariaLabelText()}
        id={`${this.props.option.value}${this.props.uniqueId}`}
        className={`checkbox-multi-select btn ${
          this.shouldDisabled() ? " disabled" : " "
        }`}
        onClick={e => this.handleCheckBoxChange(e, this.props.option)}
        onKeyDown={e => {
          this.handleKeyDown(e);
        }}
      >
        <i
          style={{ marginRight: "4px" }}
          className={
            this.props.option.selected
              ? "fa fa-check-square-o"
              : "fa fa-square-o"
          }
          aria-hidden="true"
        />
        {"  "}
        {this.props.option.label}
      </label>
    );
  }
}
