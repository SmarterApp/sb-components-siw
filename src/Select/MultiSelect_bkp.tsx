import * as React from "react";
import "../Assets/Styles/multi-select.less";

export interface MultiSelectProps {
  message: string;
}

export interface MultiSelectState {
  isDisplayOn: boolean;
}

export class MultiSelect extends React.Component<
  MultiSelectProps,
  MultiSelectState
> {
  constructor(props: MultiSelectProps) {
    super(props);
    this.state = {
      isDisplayOn: false
    };
  }

  showCheckboxes = () => {
    var checkboxes = document.getElementById("checkBoxes");
    const isDisplayOn = this.state.isDisplayOn === true ? false : true;
    this.setState({ isDisplayOn: isDisplayOn });
  };

  shiftFocus = (e: React.KeyboardEvent<HTMLElement>, idx: number) => {
    if (e.keyCode == 40) {
      idx = idx + 1;
      const ids = idx.toString();
      const dom = document.getElementById(ids);
      if (dom !== null) {
        dom.focus();
      }
    }
    //down
    if (e.keyCode == 38) {
      idx = idx - 1;
      const ids = idx.toString();
      const dom = document.getElementById(ids);
      if (dom !== null) {
        dom.focus();
      }
    }
  };

  handleFocus = () => {
    document.getElementById("1")!.focus();
  };

  renderSelectOption = () => {};

  renderDropDown = () => {
    const displayStyle = this.state.isDisplayOn === true ? "block" : "none";
    return (
      <>
        <div className="dropdown">
          <button
            className="btn btn-primary dropdown-toggle"
            type="button"
            id="dropdownMenuButton"
            data-toggle="dropdown"
            aria-haspopup="true"
            aria-expanded="false"
            onClick={this.handleFocus}
          >
            Click me
          </button>
          <form className="dropdown-menu" aria-labelledby="dropdownMenuButton">
            <label className="dropdown-item" htmlFor="1">
              <input
                tabIndex={0}
                onKeyDown={e => this.shiftFocus(e, 1)}
                type="checkbox"
                aria-aria-label="checkbox 1"
                name=""
                value="one"
                id="1"
              />First checkbox
            </label>
            <label
              className="dropdown-item"
              htmlFor="2"
              aria-aria-label="checkbox 1"
            >
              <input
                tabIndex={0}
                onKeyDown={e => this.shiftFocus(e, 2)}
                aria-aria-label="checkbox 2"
                type="checkbox"
                name=""
                value="two"
                id="2"
              />Second checkbox
            </label>
            <label className="dropdown-item" htmlFor="3">
              <input
                tabIndex={0}
                onKeyDown={e => this.shiftFocus(e, 3)}
                aria-aria-label="checkbox 3"
                type="checkbox"
                name=""
                value="three"
                id="3"
              />Third checkbox
            </label>
          </form>
        </div>
      </>
    );
  };

  render() {
    return <>{this.renderDropDown()}</>;
  }
}
