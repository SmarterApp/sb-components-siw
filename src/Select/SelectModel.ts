import { dir } from "console";

export interface MultiSelectValue {
  value: string;
  label: string;
  selected: boolean;
  disabled?: boolean;
}

export const UP_KEY = 38;
export const DOWN_KEY = 40;

export interface MultiSelectToggle {
  selectedItemsCount: number;
  isAllSelected: boolean;
}

export const multiSelectOptions: MultiSelectValue[] = [
  {
    value: "tb_in",
    label: "EBAE Contracted",
    disabled: true,
    selected: false
  },
  {
    value: "xy_zb",
    label: "EBAE with Nemeth Uncontracted",
    disabled: false,
    selected: false
  },
  {
    value: "op_gh",
    label: "EBAE with Nemeth Contracted",
    disabled: true,
    selected: false
  },
  {
    value: "xy_zb_2",
    label: "EBAE with Nemeth Uncontracted",
    disabled: false,
    selected: false
  }
];

export const getItemIndexInDirection = (
  itemArray: MultiSelectValue[],
  currentIndex: number,
  direction: string
) => {
  // let currentIndex = getIndexOfItemInArray(itemArray, currentIndexValue);
  if (currentIndex !== -1 && direction === "NEXT") {
    currentIndex += 1;
    while (currentIndex < itemArray.length) {
      if (
        itemArray[currentIndex].disabled !== undefined &&
        itemArray[currentIndex].disabled !== true
      ) {
        return currentIndex;
      }
      currentIndex++;
    }
  }
  if (currentIndex !== -1 && direction === "PREVIOUS") {
    currentIndex -= 1;
    while (currentIndex >= 0) {
      if (
        itemArray[currentIndex].disabled !== undefined &&
        itemArray[currentIndex].disabled !== true
      ) {
        return currentIndex;
      }
      currentIndex--;
    }
  }
};

export const getIndexOfItemInArray = (
  itemArray: MultiSelectValue[],
  value: string
) => {
  for (let i = 0; i < itemArray.length; i++) {
    if (itemArray[i].value === value) {
      return i;
    }
  }
  return -1;
};

export const getFirstEnabledItem = (itemArray: MultiSelectValue[]) => {
  for (let i = 0; i < itemArray.length; i++) {
    if (
      itemArray[i].disabled !== undefined &&
      itemArray[i].disabled === false
    ) {
      return i;
    }
  }
  return -1;
};

export const getButtonTextDetails = (
  itemArray: MultiSelectValue[]
): MultiSelectToggle => {
  let selectedItemsCount: number = 0;
  let isAllSelected: boolean = true;
  itemArray.forEach(item => {
    if (item.disabled !== undefined && item.disabled === false) {
      if (item.selected) {
        selectedItemsCount++;
      } else {
        isAllSelected = false;
      }
    }
  });
  return { isAllSelected, selectedItemsCount };
};

export const getFirstSelectedItem = (itemArray: MultiSelectValue[]) => {
  for (let i = 0; i < itemArray.length; i++) {
    if (
      itemArray[i].disabled !== undefined &&
      itemArray[i].disabled === false &&
      itemArray[i].selected === true
    ) {
      return itemArray[i].label;
    }
  }
  return null;
};
