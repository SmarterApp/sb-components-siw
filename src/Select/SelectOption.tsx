import * as React from "react";

/**
 * Option tag props
 * @interface SelectOptionProps
 */
export interface SelectOptionProps {
  selected: boolean;
  label: string;
  value: string;
  disabled?: boolean;
}

/**
 * Creates A11y friendly Option tag used in select tags
 * @param {SelectionOptionProps} props selection option props
 * @returns JSX React Component
 */
// tslint:disable-next-line:variable-name
export const SelectOption: React.SFC<SelectOptionProps> = props => {
  return (
    <option
      aria-selected={props.selected}
      value={props.value}
      label={props.label}
      disabled={props.disabled}
    >
      {props.label}
    </option>
  );
};
