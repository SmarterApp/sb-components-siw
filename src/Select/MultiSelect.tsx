import * as React from "react";
import "../Assets/Styles/multi-select.less";
import { MultiSelectValue } from "@src/index";
import { CheckBox } from "./CheckBox";

import {
  UP_KEY,
  DOWN_KEY,
  getItemIndexInDirection,
  getFirstEnabledItem,
  getFirstSelectedItem,
  getButtonTextDetails
} from "./SelectModel";
import ReactDOM = require("react-dom");

export interface MultiSelectProps {
  options: MultiSelectValue[];
  onChange: (v: MultiSelectValue[]) => void;
  uniqueId: number;
}

export interface MultiSelectState {
  toggleChange: number;
}

export class MultiSelect extends React.Component<
  MultiSelectProps,
  MultiSelectState
> {
  isOpenVar: boolean;
  wrapperRef: HTMLDivElement;
  setWrapperRef: any;
  node: React.RefObject<HTMLDivElement>;
  constructor(props: MultiSelectProps) {
    super(props);
    this.setWrapperRef = (element: HTMLDivElement) => {
      this.wrapperRef = element;
    };
    // this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.state = {
      toggleChange: 0
    };
    this.isOpenVar = false;
  }

  componentDidMount() {
    document.addEventListener("click", this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.handleClickOutside);
  }

  handleClickOutside(event: { target: any }) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      if (this.isOpenVar) {
        this.isOpenVar = false;
        this.setState({ toggleChange: this.state.toggleChange === 0 ? 1 : 0 });
      }
    }
  }

  onKeyPressOnButton = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.keyCode === 9) {
      this.isOpenVar = false;
      this.setState({ toggleChange: this.state.toggleChange === 0 ? 1 : 0 });
    }
  };

  shiftFocus = (
    e: React.KeyboardEvent<HTMLElement>,
    currentFocusedElement: number
  ) => {
    if (e.keyCode == UP_KEY) {
      e.preventDefault();
      const nextFocusIndex = getItemIndexInDirection(
        this.props.options,
        currentFocusedElement,
        "PREVIOUS"
      );
      if (nextFocusIndex !== undefined) {
        const dom = document.getElementById(
          this.props.options[nextFocusIndex].value +
            this.props.uniqueId.toString()
        );
        if (dom !== null) {
          window.setTimeout(function() {
            dom.focus();
          }, 0);
        }
      }
    }
    //down
    if (e.keyCode == DOWN_KEY) {
      e.preventDefault();
      const nextFocusIndex = getItemIndexInDirection(
        this.props.options,
        currentFocusedElement,
        "NEXT"
      );
      if (nextFocusIndex !== undefined) {
        const dom = document.getElementById(
          this.props.options[nextFocusIndex].value +
            this.props.uniqueId.toString()
        );
        if (dom !== null) {
          window.setTimeout(function() {
            dom.focus();
          }, 0);
        }
      }
    }

    if (e.keyCode === 9) {
      this.isOpenVar = false;
      this.setState({ toggleChange: this.state.toggleChange === 0 ? 1 : 0 });
    }

    if (e.keyCode === 13 || e.keyCode === 32) {
      e.preventDefault();
      const currentFocus = document.activeElement;
      if (currentFocus !== null)
        document.getElementById(currentFocus.id)!.click();
    }
  };

  handleFocus = () => {
    const isOpen = this.isOpenVar;
    if (isOpen) {
      this.isOpenVar = false;
      this.setState({ toggleChange: this.state.toggleChange === 0 ? 1 : 0 });
    } else {
      this.isOpenVar = true;
      this.setState(
        { toggleChange: this.state.toggleChange === 0 ? 1 : 0 },
        () => {
          this.focusFirstOption();
        }
      );
    }
  };

  focusFirstOption = () => {
    const firstFocusableIndex = getFirstEnabledItem(this.props.options);
    if (
      firstFocusableIndex !== -1 &&
      firstFocusableIndex < this.props.options.length
    ) {
      const elementId =
        this.props.options[firstFocusableIndex].value +
        this.props.uniqueId.toString();
      const element = document.getElementById(elementId);
      document.getElementById(elementId)!.focus();
    }
  };

  handleValueChange = (option: MultiSelectValue) => {
    let changeOptionsArray: MultiSelectValue[] = [];
    changeOptionsArray.push(option);
    this.props.onChange(changeOptionsArray);
    this.setState({ toggleChange: this.state.toggleChange === 0 ? 1 : 0 });
  };

  renderSelectOption = (option: MultiSelectValue, index: number) => {
    return (
      <label
        className={`multiselect-option-label ${
          option.disabled && option.disabled === true ? "disabled " : " "
        }`}
        onKeyDown={e => this.shiftFocus(e, index)}
      >
        <CheckBox
          option={option}
          key={index}
          onValueChange={this.handleValueChange}
          uniqueId={this.props.uniqueId}
        />

        {/* {"  "}
        {option.label} */}
      </label>
    );
  };

  renderSelectMultiOptionsMenu = (optionsList: MultiSelectValue[]) => {
    const multiSelectOptions = optionsList.map((option, index) => (
      // <label
      //   className={
      //     option.disabled && option.disabled === true ? "disabled " : " "
      //   }
      // >
      <li>{this.renderSelectOption(option, index)}</li>
      // </label>
    ));
    return multiSelectOptions;
  };

  renderDropDown = () => {
    //Get ready with text to be displayed in toggle button
    let buttonText: string = this.getButtonText();

    return (
      <>
        <div
          className={`dropdown ${this.isOpenVar ? "open" : ""}`}
          ref={node => this.setWrapperRef(node)}
        >
          <button
            className="btn btn-default dropdown-toggle dropdown-toggle-btn"
            type="button"
            id={"dropdownMenuButton" + this.props.uniqueId}
            data-toggle="dropdown"
            aria-haspopup="true"
            onClick={this.handleFocus}
            onKeyDown={e => {
              this.onKeyPressOnButton(e);
            }}
            tabIndex={0}
          >
            {buttonText} <i className="fa fa-angle-down" aria-hidden="true" />
          </button>
          <form
            className="dropdown-menu"
            aria-labelledby={"dropdownMenuButton" + this.props.uniqueId}
          >
            <ul className="">
              {this.renderSelectMultiOptionsMenu(this.props.options)}
            </ul>
          </form>
        </div>
      </>
    );
  };

  private getButtonText() {
    const buttonDetails = getButtonTextDetails(this.props.options);
    let buttonText: string;
    if (buttonDetails.selectedItemsCount === 0) {
      buttonText = "None selected";
    } else if (buttonDetails.isAllSelected) {
      buttonText = "All selected (" + buttonDetails.selectedItemsCount + ")";
    } else if (buttonDetails.selectedItemsCount === 1) {
      const firstSelectedItem = getFirstSelectedItem(this.props.options);
      buttonText =
        firstSelectedItem !== null ? firstSelectedItem : "1 selected";
      if (buttonText.length > 10) {
        buttonText = buttonText.replace(
          buttonText.substring(12, buttonText.length),
          " ..."
        );
      }
    } else {
      buttonText = buttonDetails.selectedItemsCount + " selected";
    }
    return buttonText;
  }

  render() {
    return <>{this.renderDropDown()}</>;
  }
}
