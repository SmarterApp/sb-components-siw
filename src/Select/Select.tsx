import * as React from "react";
import { SelectOption, SelectOptionProps } from "./SelectOption";
import { ItemCard } from "@src/ItemCard/ItemCard";
import { ItemCardModel } from "@src/ItemCard/ItemCardModels";

/**
 * Select tag properties
 * @interface SelectProps
 */
export interface SelectProps {
  onChange: (val: string, item?: ItemCardModel) => void;
  // onChangeWithItem: (item: ItemCard, val: string) => void;
  label: string;
  selected: string;
  disabled?: boolean;
  options: SelectOptionProps[];
  className?: string;
  labelClass?: string;
  wrapperClass?: string;
  tabIndex?: number;
  forItem?: ItemCardModel;
}

/**
 * Creates A11y friendly Select tag element
 * @param {SelectProps} props select props
 * @returns JSX React Component
 */
// tslint:disable-next-line:variable-name
export const Select: React.SFC<SelectProps> = props => {
  const selectOptions = props.options.map(item => (
    <SelectOption {...item} key={item.value} />
  ));

  return (
    <div className={props.wrapperClass || ""}>
      <label htmlFor={props.label}>
        <span className={props.labelClass || ""}>{props.label}</span>
      </label>
      <select
        className={`form-control ${props.className || ""}`}
        onChange={e =>
          props.onChange(
            e.target.value,
            props.forItem !== undefined ? props.forItem : undefined
          )
        }
        value={props.selected}
        id={props.label}
        aria-label={props.label}
        disabled={props.disabled}
        tabIndex={props.tabIndex}
      >
        {selectOptions}
      </select>
    </div>
  );
};
