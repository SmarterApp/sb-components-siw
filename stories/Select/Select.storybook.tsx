import * as React from "react";
import { storiesOf } from "@storybook/react";
import { MultiSelect } from "../../src/Select/MultiSelect";
import { centerDecorator } from "../CenterDecorator";
import {
  multiSelectOptions,
  MultiSelectValue
} from "../../src/Select/SelectModel";

const handleValueChange = (v: MultiSelectValue[]) => {
  console.log("chnage in storybook", v[0].value);
};

storiesOf("Select", module)
  .addDecorator(centerDecorator)
  .add("MultiSelect Dropdown", () => (
    <>
      <MultiSelect
        options={multiSelectOptions}
        onChange={handleValueChange}
        uniqueId={6000}
      />
    </>
    // <MultiSelect options={multiSelectOptions} onChange={handleValueChange} uniqueId={6010}/></>
  ));
