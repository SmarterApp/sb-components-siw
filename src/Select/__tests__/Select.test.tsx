import * as React from "react";
import { Select } from "../Select";
import { SelectOptionProps } from "../SelectOption";
import { shallow } from "enzyme";

const onChange = jest.fn();
const options: SelectOptionProps[] = [
  {
    selected: false,
    label: "one",
    value: "one"
  },
  {
    selected: false,
    label: "two",
    value: "two"
  }
];

describe("Select", () => {
  const wrapper = shallow(
    <Select
      onChange={onChange}
      label="select"
      selected="default"
      options={options}
    />
  );

  it("handles on change event", () => {
    expect(wrapper).toMatchSnapshot();
    wrapper
      .findWhere(node => node.type() === "select")
      .simulate("change", { target: { value: "one" } });
    expect(onChange).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });
});
