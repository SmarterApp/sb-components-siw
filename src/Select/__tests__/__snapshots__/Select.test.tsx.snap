// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select handles on change event 1`] = `
<div
  className=""
>
  <label
    htmlFor="select"
  >
    <span
      className=""
    >
      select
    </span>
  </label>
  <select
    aria-label="select"
    className="form-control "
    id="select"
    onChange={[Function]}
    value="default"
  >
    <Component
      key="one"
      label="one"
      selected={false}
      value="one"
    />
    <Component
      key="two"
      label="two"
      selected={false}
      value="two"
    />
  </select>
</div>
`;

exports[`Select handles on change event 2`] = `
<div
  className=""
>
  <label
    htmlFor="select"
  >
    <span
      className=""
    >
      select
    </span>
  </label>
  <select
    aria-label="select"
    className="form-control "
    id="select"
    onChange={[Function]}
    value="default"
  >
    <Component
      key="one"
      label="one"
      selected={false}
      value="one"
    />
    <Component
      key="two"
      label="two"
      selected={false}
      value="two"
    />
  </select>
</div>
`;
